File: api/src/main/java/fr/huiitre/tools/modules/core/feedback/api/FeedbackController.java
package fr.huiitre.tools.modules.core.feedback.api;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import fr.huiitre.tools.modules.core.feedback.application.usecase.BatchDeleteFeedbacksUseCase;
import fr.huiitre.tools.modules.core.feedback.application.usecase.BatchUpdateFeedbacksReadStatusUseCase;
import fr.huiitre.tools.modules.core.feedback.application.usecase.CreateFeedbackUseCase;
import fr.huiitre.tools.modules.core.feedback.application.usecase.GetAllFeedbacksUseCase;
import fr.huiitre.tools.modules.core.feedback.infrastructure.FeedbackEntity;

@RestController
@RequestMapping("/feedbacks")
public class FeedbackController {

    private final CreateFeedbackUseCase createFeedbackUseCase;
    private final GetAllFeedbacksUseCase getAllFeedbacksUseCase;
    private final BatchDeleteFeedbacksUseCase batchDeleteFeedbacksUseCase;
    private final BatchUpdateFeedbacksReadStatusUseCase batchUpdateFeedbacksReadStatusUseCase;

    public FeedbackController(
            CreateFeedbackUseCase createFeedbackUseCase,
            GetAllFeedbacksUseCase getAllFeedbacksUseCase,
            BatchDeleteFeedbacksUseCase batchDeleteFeedbacksUseCase,
            BatchUpdateFeedbacksReadStatusUseCase batchUpdateFeedbacksReadStatusUseCase) {
        this.createFeedbackUseCase = createFeedbackUseCase;
        this.getAllFeedbacksUseCase = getAllFeedbacksUseCase;
        this.batchDeleteFeedbacksUseCase = batchDeleteFeedbacksUseCase;
        this.batchUpdateFeedbacksReadStatusUseCase = batchUpdateFeedbacksReadStatusUseCase;
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public void createFeedback(@RequestBody CreateFeedbackRequest request) {
        createFeedbackUseCase.execute(request.message());
    }

    @GetMapping("/admin")
    public List<FeedbackEntity> getAllFeedbacks() {
        return getAllFeedbacksUseCase.execute();
    }

    @DeleteMapping("/admin")
    public void deleteFeedbacks(@RequestBody BatchDeleteRequest request) {
        batchDeleteFeedbacksUseCase.execute(request.ids());
    }

    @PatchMapping("/admin/read-status")
    public void updateReadStatus(@RequestBody UpdateReadStatusRequest request) {
        batchUpdateFeedbacksReadStatusUseCase.execute(request.ids(), request.isRead());
    }

    // DTOs
    public record CreateFeedbackRequest(String message) {}
    public record BatchDeleteRequest(List<Long> ids) {}
    public record UpdateReadStatusRequest(List<Long> ids, boolean isRead) {}
}

File: api/Modules/Core/Feedback/Api/FeedbackController.cs
using Microsoft.AspNetCore.Mvc;
using Tools.Api.Modules.Core.Feedback.Application;
using Tools.Api.Modules.Core.Feedback.Application.Dto;
using Tools.Api.Modules.Core.Feedback.Application.Usecases;

namespace Tools.Api.Modules.Core.Feedback.Api;


[ApiController]
[Route("feedbacks")]
public class FeedbackController : ControllerBase
{
    [HttpGet("admin")]
    public Task<List<FeedbackDto>> GetAllFeedbacks(
        [FromServices] GetAllFeedbacksUseCase getAllFeedbacksUseCase
    )
    {
        return getAllFeedbacksUseCase.Execute();
    }

    [HttpPost]
    public async Task<ActionResult> CreateFeedback(
        [FromBody] CreateFeedbackRequest request,
        [FromServices] CreateFeedbackUseCase createFeedbackUseCase
    )
    {
        await createFeedbackUseCase.Execute(request.Message);
        return Created();
    }

    [HttpDelete("admin")]
    public async Task<ActionResult> DeleteFeedbacks(
        [FromBody] BatchDeleteRequest request,
        [FromServices] DeleteFeedbacksUseCase deleteFeedbacksUseCase
    )
    {
        await deleteFeedbacksUseCase.Execute(request.FeedbackIds);
        return NoContent();
    }

    [HttpPatch("admin/read-status")]
    public async Task<ActionResult> UpdateReadStatus(
        [FromBody] UpdateReadStatusRequest request,
        [FromServices] UpdateFeedbackReadStatusUseCase updateFeedbackReadStatusUseCase
    )
    {
        await updateFeedbackReadStatusUseCase.Execute(request.Ids, request.IsRead);
        return NoContent();
    }
}

public sealed record CreateFeedbackRequest(string Message);
public sealed record BatchDeleteRequest(List<long> FeedbackIds);
public sealed record UpdateReadStatusRequest(List<long> Ids, bool IsRead);